### ISSUE TITLE
NaN for loss values and val_loss values

### ISSUE THREAD

**Issue by hacor** [eliranwong/MultiAMDGPU_AIDev_Ubuntu#5]
Hi @eliranwong 

First of all, thanks for this nice repo with info on how to install and use multiple Radeon GPUs. I had a question about my current setup:
- Dual Xeon Hexacore
- 96GB RAM
- 3x AMD Radeon Pro W7800
- Ubuntu 22.04
- ROCm 6.3.2 as stated in your docs
- I use it for tensorflow machine learning

All runs well, but when I try to run my tensorflow script:

```python
import os
import tensorflow as tf
import numpy as np
from tensorflow.keras import Input
from tensorflow.keras.models import Model
from tensorflow.keras.layers import Dense, Dropout, Flatten, Normalization
from tensorflow.keras.callbacks import ReduceLROnPlateau, EarlyStopping
from pkl_file_loader import TimeSeriesLoader

# Enable memory growth on GPUs
physical_devices = tf.config.list_physical_devices('GPU')
for gpu_instance in physical_devices:
    tf.config.experimental.set_memory_growth(gpu_instance, True)

# Multi-GPU Strategy
strategy = tf.distribute.MirroredStrategy(cross_device_ops=tf.distribute.HierarchicalCopyAllReduce())
print(f"🧠 Number of GPUs detected: {strategy.num_replicas_in_sync}")

# Define input shape
history_input_shape = (21*96, 1)

# Build model inside strategy scope
with strategy.scope():
    history_input = Input(shape=history_input_shape, name='history_input')
    x = Flatten()(history_input)
    x = Normalization()(x)
    x = Dense(512, activation='relu')(x)
    x = Dropout(0.15)(x)
    x = Dense(512, activation='relu')(x)
    x = Dropout(0.15)(x)
    x = Dense(512, activation='relu')(x)
    x = Dropout(0.15)(x)
    output = Dense(96, activation='linear', name="building_output")(x)

    model = Model(inputs=history_input, outputs=output)

    initial_learning_rate = 0.0001
    final_learning_rate = 1e-7
    reduce_lr = ReduceLROnPlateau(monitor='val_loss', factor=0.1, patience=10, min_lr=final_learning_rate)
    early_stopping = EarlyStopping(monitor='val_loss', patience=20, min_delta=1e-4, restore_best_weights=True)

    model.compile(
        optimizer=tf.keras.optimizers.Adam(learning_rate=initial_learning_rate, clipnorm=1.0),
        loss=tf.keras.losses.Huber(delta=1.0),
        metrics=['mae']
    )

model.summary()

# Load dataset
tss = TimeSeriesLoader()
train_x, train_y = tss.get_chunk(0)

# Make sure input dimensions are correct
train_x = np.array(train_x, dtype=np.float32)
train_y = np.array(train_y, dtype=np.float32)
if train_x.ndim == 2:
    train_x = train_x[..., np.newaxis]

# Prepare datasets
global_batch_size = 24 * strategy.num_replicas_in_sync

full_dataset = tf.data.Dataset.from_tensor_slices((train_x, train_y)).shuffle(buffer_size=10000)
val_size = int(0.2 * len(train_x))
train_dataset = full_dataset.skip(val_size)
val_dataset = full_dataset.take(val_size)

train_dataset = train_dataset.batch(global_batch_size).prefetch(tf.data.AUTOTUNE)
val_dataset = val_dataset.batch(global_batch_size).prefetch(tf.data.AUTOTUNE)

# Train the model
history = model.fit(
    train_dataset,
    validation_data=val_dataset,
    epochs=200,
    callbacks=[reduce_lr, early_stopping]
)

# Save the model
model.save('build/usage_model.keras')
```

All `loss` and `val_loss` values go to `NaN`, but when I run the script without the multiple GPU support `tf.distribute.MirroredStrategy(cross_device_ops=tf.distribute.HierarchicalCopyAllReduce())` it works perfectly.

These are the logs:
```bash
(.venv) (base) hacor@hacor-linux-box:~/tensorflow-models/house-usage$ python 02_train_model_6.py
2025-05-05 17:48:15.713860: I tensorflow/core/platform/cpu_feature_guard.cc:210] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.
To enable the following instructions: SSE3 SSE4.1 SSE4.2 AVX AVX2 FMA, in other operations, rebuild TensorFlow with the appropriate compiler flags.
2025-05-05 17:48:22.586549: I tensorflow/core/common_runtime/gpu/gpu_device.cc:2021] Created device /job:localhost/replica:0/task:0/device:GPU:0 with 29652 MB memory:  -> device: 0, name: AMD Radeon PRO W7800, pci bus id: 0000:83:00.0
2025-05-05 17:48:22.929800: I tensorflow/core/common_runtime/gpu/gpu_device.cc:2021] Created device /job:localhost/replica:0/task:0/device:GPU:1 with 29652 MB memory:  -> device: 1, name: AMD Radeon PRO W7800, pci bus id: 0000:08:00.0
2025-05-05 17:48:23.264598: I tensorflow/core/common_runtime/gpu/gpu_device.cc:2021] Created device /job:localhost/replica:0/task:0/device:GPU:2 with 29652 MB memory:  -> device: 2, name: AMD Radeon PRO W7800, pci bus id: 0000:05:00.0
🧠 Number of GPUs detected: 3
Model: "functional"
┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━┓
┃ Layer (type)                         ┃ Output Shape                ┃         Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━┩
│ history_input (InputLayer)           │ (None, 2016, 1)             │               0 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ flatten (Flatten)                    │ (None, 2016)                │               0 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ normalization (Normalization)        │ (None, 2016)                │           4,033 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dense (Dense)                        │ (None, 512)                 │       1,032,704 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dropout (Dropout)                    │ (None, 512)                 │               0 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dense_1 (Dense)                      │ (None, 512)                 │         262,656 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dropout_1 (Dropout)                  │ (None, 512)                 │               0 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dense_2 (Dense)                      │ (None, 512)                 │         262,656 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dropout_2 (Dropout)                  │ (None, 512)                 │               0 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ building_output (Dense)              │ (None, 96)                  │          49,248 │
└──────────────────────────────────────┴─────────────────────────────┴─────────────────┘
 Total params: 1,611,297 (6.15 MB)
 Trainable params: 1,607,264 (6.13 MB)
 Non-trainable params: 4,033 (15.76 KB)
Epoch 1/200
 692/9421 ━━━━━━━━━━━━━━━━━━━━ 7:23 51ms/step - loss: nan - mae: nan
```

I hope you could give me a clue on what's going wrong.
Thank you!

Kind Regards

Hacor

**Comment by NazimKYS** [comment]
Hi @hacor,
I’ve been looking into open issues on GitHub related to **multi-GPU training**, and I noticed that many users have reported problems when using **TensorFlow ≥ 2.16.0** and/or **Keras ≥ 3.7.0.** These versions appear to cause multi-GPU training to fail.

Based on what you described, it seems like you might be running into the same issue. A commonly suggested workaround is to downgrade **TensorFlow to 2.15.0 and Keras to 3.6.0**, which has resolved the problem for many users.

I recommend giving that a try and letting me know if it solves your issue.

**Comment by hacor** [comment]
Nice! thanks!
Downgrading to `tensorflow==2.15.0` fixed it!
Closing issue, hoping it can help others
